src/toil/job.py

```diff
@@ -59,246 +59,897 @@
+                self.command = self.checkpoint
+
+            jobStore.update(self) # Update immediately to ensure that checkpoint
+            # is made before deleting any remaining successors
+
+            if self.childIDs or self.followOnIDs or self.serviceTree:
+                # If the subtree of successors is not complete restart everything
+                logger.debug("Checkpoint job has unfinished successor jobs, deleting children: %s, followOns: %s, services: %s " %
+                             (self.childIDs, self.followOnIDs, self.serviceTree.keys()))
+                # Delete everything on the stack, as these represent successors to clean
+                # up as we restart the queue
+                def recursiveDelete(jobDesc):
+                    # Recursive walk the stack to delete all remaining jobs
+                    for otherJobID in jobDesc.successorsAndServiceHosts():
+                        if jobStore.exists(otherJobID):
+                            recursiveDelete(jobStore.load(otherJobID))
+                        else:
+                            logger.debug("Job %s has already been deleted", otherJobID)
+                    if jobDesc.jobStoreID != self.jobStoreID:
+                        # Delete everything under us except us.
+                        logger.debug("Checkpoint is deleting old successor job: %s", jobDesc.jobStoreID)
+                        jobStore.delete(jobDesc.jobStoreID)
+                        successorsDeleted.append(jobDesc.jobStoreID)
+                recursiveDelete(self)
+                
+                # Cut links to the jobs we deleted.
+                self.clearSuccessorsAndServiceHosts()
 
-    @classmethod
-    def fromJob(cls, job, command, predecessorNumber):
-        """
-        Build a job node from a job object
-        :param toil.job.Job job: the job object to be transformed into a job node
-        :param str command: the JobNode's command
-        :param int predecessorNumber: the number of predecessors that must finish
-            successfully before the job can be scheduled
-        :return: a JobNode object representing the job object parameter
-        :rtype: toil.job.JobNode
-        """
-        return cls(jobStoreID=None,
-                   requirements=job._requirements,
-                   command=command,
-                   jobName=job.jobName,
-                   unitName=job.unitName,
-                   displayName=job.displayName,
-                   predecessorNumber=predecessorNumber)
-
-class Job(BaseJob):
+                # Update again to commit the removal of successors.
+                jobStore.update(self)
+        return successorsDeleted
+
+class Job:
     """
     Class represents a unit of work in toil.
     """
     def __init__(self, memory=None, cores=None, disk=None, preemptable=None,
-                       unitName=None, checkpoint=False, displayName=None):
+                       unitName='', checkpoint=False, displayName='',
```

Comment:
Could typing be added here?

Reply:
I think these all have types documented already.